simulation/environment/CacheAPI.py

```diff
@@ -0,0 +1,503 @@
+                path_points = path_points + polyline_coords
+
+        print("Route has been successfully retrieved!\n")
+
+    else:
+        print(f"No route was found: {response['status']}")
+        print(f"Error Message: {response['error_message']}")
+
+    route = np.array(path_points)
+
+    # Removes duplicate coordinates to prevent gradient calculation errors
+    if route.size != 0:
+        duplicate_coordinate_indices = np.where((np.diff(route[:, 0]) == 0)) and np.where(
+            (np.diff(route[:, 1]) == 0))
+        route = np.delete(route, duplicate_coordinate_indices, axis=0)
+
+    return origin_coord, dest_coord, route, waypoints
+
+
+def calculate_path_elevations(coords):
+    """
+
+    Returns the elevations of every coordinate in the array of coordinates passed in as a coordinate
+    See Error Message Interpretations: https://developers.google.com/maps/documentation/elevation/overview
+
+    :param np.ndarray coords: A NumPy array [n][latitude, longitude]
+    :returns: A NumPy array [n][elevation] in metres
+    :rtype: np.ndarray
+
+    """
+
+    # construct URL
+    url_head = 'https://maps.googleapis.com/maps/api/elevation/json?locations='
+
+    location_strings = []
+    locations = ""
+
+    for coord in coords:
+
+        locations = locations + f"{coord[0]},{coord[1]}|"
+
+        if len(locations) > 8000:
+            location_strings.append(locations[:-1])
+            locations = ""
+
+    if len(locations) != 0:
+        location_strings.append(locations[:-1])
+
+    url_tail = f"&key={os.environ['GOOGLE_MAPS_API_KEY']}"
+
+    # Get elevations
+    elevations = np.zeros(len(coords))
+
+    i = 0
+    with tqdm(total=len(location_strings), file=sys.stdout, desc="Acquiring Elevation Data") as pbar:
+        for location_string in location_strings:
+            url = url_head + location_string + url_tail
+
+            r = requests.get(url)
+            response = json.loads(r.text)
+            pbar.update(1)
+
+            if response['status'] == "OK":
+                for result in response['results']:
+                    elevations[i] = result['elevation']
+                    i = i + 1
+
+            elif response['status'] == "INVALID_REQUEST":
+                sys.stderr.write("Error: Request was invalid\n")
+
+            elif response['status'] == "OVER_DAILY_LIMIT":
+                sys.stderr.write(
+                    "Error: Possible causes - API key is missing or invalid, billing has not been enabled,"
+                    " a self-imposed usage cap has been exceeded, or the provided payment method is no longer "
+                    " valid. \n")
+
+            elif response['status'] == "OVER_QUERY_LIMIT":
+                sys.stderr.write("Error: Requester has exceeded quota\n")
+
+            elif response['status'] == "REQUEST_DENIED":
+                sys.stderr.write("Error: API could not complete the request\n")
+
+    return elevations
+
+
+def calculate_time_zones(coords, race):
```

Comment:
`GIS` needs this function on line 96, consider moving it to `helpers` so both of them can access it